src/test/java/org/junitpioneer/jupiter/report/PioneerReportListenerTest.java

```diff
@@ -0,0 +1,158 @@
+
+	@Test
+	void emptyReport() {
+		listener.testPlanExecutionStarted(testPlan);
+		listener.testPlanExecutionFinished(testPlan);
+
+		// Verify result
+		PioneerReport report = listener.getPioneerReport();
+		assertThat(report.getProperties()).isNull();
+		assertThat(report.getTestcontainer().isEmpty()).isTrue();
+		assertThat(report.getTestcase().isEmpty()).isTrue();
+	}
+
+	@Test
+	void reportingCorrectCounts() {
+
+		ReportEntry succContainerEntry = ReportEntry.from("key: succContainer", "value: succContainer");
+		ReportEntry succTestEntry = ReportEntry.from("key: succTest", "value: succTest");
+
+		TestIdentifier successfulContainer = createContainerIdentifier("c1");
+		TestIdentifier failedContainer = createContainerIdentifier("c2");
+		TestIdentifier abortedContainer = createContainerIdentifier("c3");
+		TestIdentifier skippedContainer = createContainerIdentifier("c4");
+
+		TestIdentifier successfulTest = createTestIdentifier("t1");
+		TestIdentifier failedTest = createTestIdentifier("t2");
+		TestIdentifier abortedTest = createTestIdentifier("t3");
+		TestIdentifier skippedTest = createTestIdentifier("t4");
+
+		listener.testPlanExecutionStarted(testPlan);
+
+		listener.executionSkipped(skippedContainer, "skipped");
+		listener.executionSkipped(skippedTest, "skipped");
+
+		listener.executionStarted(successfulContainer);
+		listener.executionFinished(successfulContainer, TestExecutionResult.successful());
+
+		listener.reportingEntryPublished(successfulContainer, succContainerEntry);
+
+		listener.executionStarted(successfulTest);
+		listener.executionFinished(successfulTest, TestExecutionResult.successful());
+
+		listener.reportingEntryPublished(successfulTest, succTestEntry);
+
+		listener.executionStarted(failedContainer);
+		listener.executionFinished(failedContainer, TestExecutionResult.failed(new RuntimeException("failed")));
+
+		listener.executionStarted(failedTest);
+		listener.executionFinished(failedTest, TestExecutionResult.failed(new RuntimeException("failed")));
+
+		listener.executionStarted(abortedContainer);
+		listener.executionFinished(abortedContainer, TestExecutionResult.aborted(new RuntimeException("aborted")));
+
+		listener.executionStarted(abortedTest);
+		listener.executionFinished(abortedTest, TestExecutionResult.aborted(new RuntimeException("aborted")));
+
+		listener.testPlanExecutionFinished(testPlan);
+
+		// Verify result
+		PioneerReport report = listener.getPioneerReport();
+		assertThat(report.getProperties()).isNull();
+
+		// Verify test containers
+		assertThat(report.getTestcontainer().size()).isEqualTo(4);
+		assertThat(report.getTestcontainer().get(0).getName()).isIn("c1", "c2", "c3", "c4");
+
+		for (Testcontainer testcontainer : report.getTestcontainer()) {
+			if (testcontainer.getName().equals("c1")) {
+				Property property = testcontainer.getProperties().getProperty().get(0);
+				assertThat(property.getName()).isEqualTo("key: succContainer");
```

Comment:
Could you move the assertions out from the loop? I think they would look better.

Reply:
Also if you move them out this can be replaced with a `Stream`.

Reply:
Updated